pydantic_ai_slim/pydantic_ai/models/rate_limited.py

```diff
@@ -0,0 +1,165 @@
+from __future__ import annotations
+
+from collections.abc import AsyncIterator
+from contextlib import asynccontextmanager
+from dataclasses import dataclass
+from typing import TYPE_CHECKING
+
+from tenacity import AsyncRetrying
+
+if TYPE_CHECKING:
+    from throttled.asyncio import Throttled
+
+from . import (
+    KnownModelName,
+    Model,
+    ModelMessage,
+    ModelRequestParameters,
+    ModelResponse,
+    ModelSettings,
+    StreamedResponse,
+)
+from .wrapper import WrapperModel
+
+
+@dataclass
+class RateLimitedModel(WrapperModel):
+    """Model which wraps another model such that requests are rate limited with throttled.
+
+    If retryer is provided it also retries requests with tenacity.
+
+    Usage:
+
+    ```python
+    from tenacity import AsyncRetrying, stop_after_attempt
+    from throttled.asyncio import Throttled, rate_limiter
+    from throttled.asyncio.store import MemoryStore
+
+    from pydantic_ai import Agent
+    from pydantic_ai.models.rate_limited import RateLimitedModel
+
+    throttle = Throttled(
+        using='gcra',
+        quota=rate_limiter.per_sec(1_000, burst=1_000),
+        store=MemoryStore(),
+    )
+
+    model = RateLimitedModel(
+        'anthropic:claude-3-7-sonnet-latest',
+        limiter=throttle,
+        retryer=AsyncRetrying(stop=stop_after_attempt(3)),
+    )
+
+    agent = Agent(model=model)
+    ```
+    """
+
+    def __init__(
+        self,
+        wrapped: Model | KnownModelName,
+        limiter: Throttled | None = None,
+        retryer: AsyncRetrying | None = None,
+    ) -> None:
+        super().__init__(wrapped)
+        self.limiter = limiter
+        self.retryer = retryer
+
+    async def request(
+        self,
+        messages: list[ModelMessage],
+        model_settings: ModelSettings | None,
+        model_request_parameters: ModelRequestParameters,
+        key: str = 'default',
+        cost: int = 1,
+        timeout: float | None = 30.0,
```

Comment:
Would it make sense to move these to the initializer? Or could they be values on the limiter that's passed? Users don't call `request` and `request_stream` directly, so they wouldn't have an obvious way to set these values.